tempesta_fw/http.c

```diff
@@ -2794,25 +2838,94 @@ tfw_http_req_process(TfwConn *conn, const TfwFsmData *data)
 			       && (req->content_length != req->body.len));
 		}
 
-		/* Assign the right virtual host for current request. */
-		if ((req->vhost = tfw_http_tbl_vhost((TfwMsg *)req, &block)))
+		/*
+		 * Sticky cookie module must be used before request can reach
+		 * cache. Unauthorised clients mustn't be able to get any
+		 * resource on protected service and stress cache subsystem.
+		 * The module is also quickest way to obtain target VHost and
+		 * target backend server connection since it allows to
+		 * avoid expensive tables lookups.
+		 *
+		 * TODO: #685 Sticky cookie are to be configured per-vhost.
+		 * Http_tables used to assign target vhost can be extremely
+		 * long, while the client may use just a few cookie variants.
+		 * Even if different sticky cookies are used for each vhost,
+		 * the sticky module should be faster than
+		 * tfw_http_tbl_vhost().
+		 *
+		 * TODO: #1043 sticky cookie module must assign correct vhost
+		 * when it returns TFW_HTTP_SESS_SUCCESS or
+		 * TFW_HTTP_SESS_REDIRECT_NEED;
+		 */
+		switch (tfw_http_sess_obtain(req))
+		{
+		case TFW_HTTP_SESS_SUCCESS:
+			break;
+
+		case TFW_HTTP_SESS_REDIRECT_NEED:
+			/* Response is built and stored in @req->resp. */
+			break;
+
+		case TFW_HTTP_SESS_VIOLATE:
+			TFW_INC_STAT_BH(clnt.msgs_filtout);
+			tfw_http_req_parse_block(
+				req, 503,
+				"request dropped: sticky cookie challenge was "
+				"failed");
+			return TFW_BLOCK;
+
+		case TFW_HTTP_SESS_JS_NOT_SUPPORTED:
+			/*
+			 * Requested resource can't be challenged, forward all
+			 * pending responses and close the connection to allow
+			 * client to recover.
+			 */
+			TFW_INC_STAT_BH(clnt.msgs_filtout);
+			tfw_http_req_parse_block(
+				req, 503,
+				"request dropped: can't send JS challenge");
+			return TFW_BLOCK;
+
+		default:
+			TFW_INC_STAT_BH(clnt.msgs_otherr);
+			tfw_http_req_parse_block(
+				req, 500,
+				"request dropped: internal error in Sticky "
+				"module");
+			return TFW_BLOCK;
+		}
+
+		/*
+		 * Assign the right virtual host for current request if it
+		 * wasn't assigned by a sticky module. VHost is not assigned
+		 * if client doesn't support cookies or sticky cookies are
+		 * disabled.
+		 *
+		 * In the same time location may differ between requests, so the
+		 * sticky module can't fill it.
+		 */
+		if (!req->vhost)
+			req->vhost = tfw_http_tbl_vhost((TfwMsg *)req, &block);
+		if (req->vhost)
 			req->location = tfw_location_match(req->vhost,
 							   &req->uri_path);
+		else if (block) {
```

Comment:
If HTTPtables didn't return a vhost, why they can't block the request, i.e. why the `else` appeared?

It'd be better to use curly braces for the `if` hading `else if` with curly braces.